thunder/core/transform_common.py

```diff
@@ -363,3 +367,101 @@ class PostOptimizationTransform(Transform, ABC):
     @abstractmethod
     def transform_trace(self, computation_trace: Trace, **kwargs):
         pass
+
+
+def functionalize_inplace_ops(computation_trace: Trace) -> list[Trace]:
+    """Functionalize in-place ops in ``computation_trace``.
+
+    This is a two-step function. The first step is to use the output of :func:`thunder.core.prims.copy_`'s
+    outputs as trace's outputs. The second step is to remove :func:`thunder.core.prims.copy_`
+    from :attr:`~thunder.core.symbol.Boundsymbol.subsymbols` if it's the last symbol.
+    """
+    from thunder.core import utils
+    from thunder.core.proxies import ProxyInterface
+    from thunder.core.symbol import VariableInterface
+    import thunder.torch
+
+    def is_inplace(bsym: BoundSymbol) -> bool:
+        if (isinstance(bsym.sym.id, str) and bsym.sym.id.endswith("_")) and bsym.subsymbols:
+            return bsym.subsymbols[-1].sym.id == prims.PrimIDs.COPY_
+        else:
+            return False
+
+    if not any(is_inplace(bsym) for bsym in computation_trace.bound_symbols):
+        return []
+
+    # Step 1: return the tensors returned from `prims.copy_` as possible not the args for clarity.
+    bsym: BoundSymbol
+    swap_map: dict[VariableInterface, ProxyInterface] = {}
+    bsyms: list[BoundSymbol] = []
+    for bsym in computation_trace.bound_symbols:
+        new_bsym = bsym.from_bsym_swap_proxies(swap_map)
+
+        # in-place ops has `prims.copy_` as the last subsymbol.
+        if not is_inplace(new_bsym):
+            bsyms.append(new_bsym)
+            continue
+
+        sub_bsyms: list[BoundSymbol] = bsym.subsymbols
+        utils.check(sub_bsyms, lambda: f"{bsym.sym.id=} expected to have subsymbols but {bsym.subsymbols=}")
+        copy_bsym = sub_bsyms[-1]
```

Comment:
Similarly, this assumption doesn't hold for batchnorm?

Reply:
But that might not be a bad thing. for what I know, we cannot handle batchnorm running stats update with functionalization anyway. So maybe we would want to keep it as-is.

Reply:
> Similarly, this assumption doesn't hold for batchnorm?

if batchnorm's `torchsymbol` exists and it doesn't have `OpTags.IN_PLACE` then it's free from the possibility of functionalization.

Reply:
I agree, `prims.copy_` from subsymbols of batch norm should not be functionalized. It's our way to insert in-place copy in a controlled setting. Also it shouldn't functionalized because the tensors that are modified in-place are inputs to the compiled function. Currently, it's achieved by not marking batchnorm's torchsymbol with the IN_PLACE tag.
Many operations can be tagged as in-place, but they should be functionalized only if they act on tensors that were produced within the compiled function, is this implemented in the current version of the pull request?

Reply:
If we already have commit 8889120 that's guarding the functionalization of inplace update on inputs, is there anything else we needed to special case for batch norm?

Reply:
I don't think so given the change of [test_batch_norm_running_stats](https://github.com/Lightning-AI/lightning-thunder/pull/584/files#diff-147ca1270ed52defb8a1ad70ef9c0fb88b989be45bdaedd938fc98befa452226)